src/utilcode/pedecoder.cpp

```diff
@@ -1790,7 +1790,96 @@ void PEDecoder::LayoutILOnly(void *base, BOOL allowFullPE) const
 
 #endif // #ifndef DACCESS_COMPILE
 
-#ifndef FEATURE_PAL
+DWORD PEDecoder::ReadResourceDictionary(DWORD rvaOfResourceSection, DWORD rva, LPCWSTR name, BOOL *pIsDictionary) const
+{
+    *pIsDictionary = FALSE;
+
+    if (!CheckRva(rva, sizeof(IMAGE_RESOURCE_DIRECTORY)))
+    {
+        return 0;
+    }
+
+    IMAGE_RESOURCE_DIRECTORY *pResourceDirectory = (IMAGE_RESOURCE_DIRECTORY *)GetRvaData(rva);
+
+    if (pResourceDirectory->MajorVersion != 4)
+        return 0;
+
+    if (pResourceDirectory->MinorVersion != 0)
+        return 0;
+
+    // Check to see if entire resource dictionary is accessible
+    if (!CheckRva(rva + sizeof(IMAGE_RESOURCE_DIRECTORY), 
+                       (sizeof(IMAGE_RESOURCE_DIRECTORY_ENTRY) * pResourceDirectory->NumberOfNamedEntries) +
+                       (sizeof(IMAGE_RESOURCE_DIRECTORY_ENTRY) * pResourceDirectory->NumberOfIdEntries)))
+    {
+        return 0;
+    }
+
+    IMAGE_RESOURCE_DIRECTORY_ENTRY* pDirectoryEntries = (IMAGE_RESOURCE_DIRECTORY_ENTRY *)GetRvaData(rva + sizeof(IMAGE_RESOURCE_DIRECTORY));
+
+    // A fast implementation of resource lookup uses a binary search, but our needs are simple, and a linear search
+    // is easier to prove correct, so do that instead.
+    DWORD iEntryCount = (DWORD)pResourceDirectory->NumberOfNamedEntries + (DWORD)pResourceDirectory->NumberOfIdEntries;
+
+    for (DWORD iEntry = 0; iEntry < iEntryCount; iEntry++)
+    {
+        BOOL foundEntry = FALSE;
+
+        if (((UINT_PTR)name) <= 0xFFFF)
+        {
+            // name is id
+            if (pDirectoryEntries[iEntry].Name == (DWORD)name)
+                foundEntry = TRUE;
+        }
+        else
+        {
+            // name is string
+            DWORD entryName = pDirectoryEntries[iEntry].Name;
+            if (!(entryName & IMAGE_RESOURCE_NAME_IS_STRING))
+                continue;
+            
+            DWORD entryNameRva = (entryName & ~IMAGE_RESOURCE_NAME_IS_STRING) + rvaOfResourceSection;
+
+            if (!CheckRva(entryNameRva, sizeof(WORD)))
+                return 0;
+
+            size_t entryNameLen = *(WORD*)GetRvaData(entryNameRva);
+            if (wcslen(name) != entryNameLen)
```

Comment:
wcslen against untrusted data. You need to inline it and adds bounds checking. As well, this seems wrong for cross-platform but I lack context -- is this native platform wchar_t-based that is often 32bit wchar_t, or redirected to always 16bit.

Reply:
It is redirected to be 16bit WCHAR on all platforms.

Reply:
Also, name is known to be of a safe length in all current uses of this function, and new uses are not expected.